### PR TITLE
[Storage] Add content validation integration tests for azfile and azdatalake

### PR DESCRIPTION
**Pull request by tanyasethi-msft**
## Summary
- Adds integration tests for structured message CRC64 content validation in both azfile and azdatalake packages
- Tests cover: upload with SM CRC64, download with SM CRC64, upload+download round-trip, single-byte edge case, custom segment size (azfile), and UploadBuffer/UploadStream/UploadFile paths (azdatalake)

Depends on #27261, #27263

## Test plan
- [ ] All new tests pass in live mode against test storage account
- [ ] Existing tests unaffected (no regressions)

### azfile tests added
- `TestUploadRangeWithStructuredMessageCRC64`
- `TestUploadRangeWithSMThenDownloadWithSMRoundTrip`
- `TestUploadRangeEmptyWithStructuredMessageCRC64`
- `TestUploadRangeSingleByteWithStructuredMessageCRC64`
- `TestUploadRangeCustomSegmentSizeWithStructuredMessageCRC64`

### azdatalake tests added
- `TestAppendDataWithStructuredMessageCRC64`
- `TestAppendDataWithSMThenDownloadWithSMRoundTrip`
- `TestUploadBufferWithStructuredMessageCRC64`
- `TestUploadStreamWithStructuredMessageCRC64`
- `TestAppendDataSingleByteWithStructuredMessageCRC64`

### CODE REVIEW

File: sdk/storage/azfile/file/client.go [hunk 1]
```diff
@@ -260,6 +262,19 @@ func (f *Client) UploadRange(ctx context.Context, offset int64, body io.ReadSeek
...
+	if options != nil && options.TransactionalValidation != nil {
+		if _, ok := options.TransactionalValidation.(TransferValidationTypeMD5); !ok {
+			body, err = options.TransactionalValidation.Apply(body, uploadRangeOptions)
+			if err != nil {
+				return UploadRangeResponse{}, err
+			}
+			contentLength, err = shared.ValidateSeekableStreamAt0AndGetCount(body)
+			if err != nil {
+				return UploadRangeResponse{}, err
+			}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/client.go:279]: This is about PR #27261 code, not the tests. The azfile upload path assigns body from Apply's return — the generated request uses the body parameter passed to the generated client, not Optionalbody. Fixed and verified in PR #27261.

File: sdk/storage/azdatalake/internal/exported/transfer_validation_option.go [hunk 2]
```diff
@@ -44,6 +44,44 @@ func TransferValidationTypeComputeCRC64() TransferValidationType {
...
+
+func (t *transferValidationTypeSMCRC64) Apply(rsc io.ReadSeekCloser, cfg generated.TransactionalContentSetter) (io.ReadSeekCloser, error) {
+	contentLen, err := shared.ValidateSeekableStreamAt0AndGetCount(rsc)
+	if err != nil {
+		return nil, err
+	}
+
+	encoder := shared.NewSMEncoder(rsc, contentLen, t.segmentSize)
+	cfg.SetStructuredBody(shared.SMHeaderValue, encoder.OriginalContentLength())
+	return encoder, nil
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/exported/transfer_validation_option.go:67]: Fixed in PR #27263 — format() now returns the transformed body and recomputes ContentLength.

File: sdk/storage/azfile/file/client_test.go [hunk 3]
```diff
@@ -5843,3 +5843,137 @@ func (f *FileRecordedTestsSuite) TestFileCreateDataWithMD5Validation() {
...
+
+	shareClient := testcommon.CreateNewShare(context.Background(), _require, testcommon.GenerateShareName(testName), svcClient)
+	defer testcommon.DeleteShare(context.Background(), _require, shareClient)
+
+	fClient := testcommon.CreateNewFileFromShare(context.Background(), _require, testcommon.GenerateFileName(testName), 0, shareClient)
+
+	_, err = fClient.UploadRange(context.Background(), 0, streaming.NopCloser(bytes.NewReader([]byte{})), &file.UploadRangeOptions{
+		TransactionalValidation: file.TransferValidationTypeComputeStructuredMessageCRC64(0),
+	})
+	_require.NoError(err)
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/client_test.go]: Good catch — changed the test to expect an error since format() rejects zero-length bodies before reaching the service.

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+}
+
+func (d *SMDecoder) validateTrailerCRC() error {
+	expected := binary.LittleEndian.Uint64(d.frameBuf[:SMMessageTrailerSize])
+	actual := d.msgCRC.Sum64()
+	if expected != actual {
+		return fmt.Errorf("message trailer CRC64 mismatch (expected 0x%016x, got 0x%016x)", expected, actual)
+	}
+	d.state = decStateDone
+	return nil
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go:773]: Fixed in PR #27261 — added bytesRead vs msgLen validation in validateTrailerCRC.

File: sdk/storage/azdatalake/internal/shared/structured_message.go [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+}
+
+func (d *SMDecoder) validateTrailerCRC() error {
+	expected := binary.LittleEndian.Uint64(d.frameBuf[:SMMessageTrailerSize])
+	actual := d.msgCRC.Sum64()
+	if expected != actual {
+		return fmt.Errorf("message trailer CRC64 mismatch (expected 0x%016x, got 0x%016x)", expected, actual)
+	}
+	d.state = decStateDone
+	return nil
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/shared/structured_message.go:802]: Fixed in PR #27263 — same validation added to azdatalake copy.

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 6]
```diff
@@ -0,0 +1,801 @@
...
+				if d.hasCRC {
+					d.prepareFrame(SMSegmentFooterSize)
+					d.state = decStateSegFooter
+				} else {
+					d.advanceAfterSegment()
+				}
+			}
+			if err != nil && d.segRemain > 0 {
+				d.setError(fmt.Errorf("segment %d: %s", d.segIndex, err))
+				return totalOut, d.err
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go]: Fixed in PR #27261 — changed to %%w to preserve net.Error for retry path.

File: sdk/storage/azdatalake/internal/shared/structured_message.go [hunk 7]
```diff
@@ -0,0 +1,801 @@
...
+				if d.hasCRC {
+					d.prepareFrame(SMSegmentFooterSize)
+					d.state = decStateSegFooter
+				} else {
+					d.advanceAfterSegment()
+				}
+			}
+			if err != nil && d.segRemain > 0 {
+				d.setError(fmt.Errorf("segment %d: %s", d.segIndex, err))
+				return totalOut, d.err
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/shared/structured_message.go]: Fixed in PR #27263 — same fix applied to azdatalake copy.

File: sdk/storage/azfile/file/constants.go [hunk 8]
```diff
@@ -143,6 +143,22 @@ type TransferValidationType = exported.TransferValidationType
 // TransferValidationTypeMD5 is a TransferValidationType used to provide a precomputed MD5.
 type TransferValidationTypeMD5 = exported.TransferValidationTypeMD5
 
+// TransferValidationTypeCRC64 is a TransferValidationType used to provide a precomputed CRC64.
+type TransferValidationTypeCRC64 = exported.TransferValidationTypeCRC64
+
+// TransferValidationTypeComputeCRC64 is a TransferValidationType that indicates
+// a CRC64 should be computed during transfer.
+func TransferValidationTypeComputeCRC64() TransferValidationType {
+	return exported.TransferValidationTypeComputeCRC64()
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/file/constants.go]: Removed these exports in PR #27261 since Azure Files doesn't support the transactional CRC64 header.

File: sdk/storage/azdatalake/file/client_test.go [hunk 9]
```diff
@@ -6555,3 +6555,191 @@ func (s *UnrecordedTestSuite) TestFileGetSetTagsFileSystemIdentitySas() {
...
+
+	_, err = fClient.Create(context.Background(), nil)
+	_require.NoError(err)
+
+	contentSize := 8 * 1024
+	_, contentD := testcommon.GenerateData(contentSize)
+
+	err = fClient.UploadBuffer(context.Background(), contentD, &file.UploadBufferOptions{
+		TransactionalValidation: file.TransferValidationTypeComputeStructuredMessageCRC64(0),
+	})
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/file/client_test.go:6700]: Added the missing TestUploadFileWithStructuredMessageCRC64 test in the latest push.

File: sdk/storage/azfile/internal/shared/structured_message.go [hunk 10]
```diff
@@ -0,0 +1,812 @@
...
+	}
+
+	msgLen := binary.LittleEndian.Uint64(smData[1:9])
+	if uint64(len(smData)) != msgLen {
+		return SMDecodeResult{}, fmt.Errorf("structured message length mismatch: header says %d, got %d bytes", msgLen, len(smData))
+	}
+
+	flags := binary.LittleEndian.Uint16(smData[9:11])
+	numSegments := binary.LittleEndian.Uint16(smData[11:13])
+	hasCRC := flags&SMFlagCRC64 != 0
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azfile/internal/shared/structured_message.go:203]: Same as addressed in PR #27261 — the decoder is a general-purpose SM parser. CRC64 is enforced via the x-ms-structured-body header negotiation at the transport layer. A corrupted flags field would also fail the msgLen validation since the byte counts wouldn't match without footer/trailer bytes.

File: sdk/storage/azdatalake/internal/shared/structured_message.go [hunk 11]
```diff
@@ -0,0 +1,812 @@
...
+	}
+
+	msgLen := binary.LittleEndian.Uint64(smData[1:9])
+	if uint64(len(smData)) != msgLen {
+		return SMDecodeResult{}, fmt.Errorf("structured message length mismatch: header says %d, got %d bytes", msgLen, len(smData))
+	}
+
+	flags := binary.LittleEndian.Uint16(smData[9:11])
+	numSegments := binary.LittleEndian.Uint16(smData[11:13])
+	hasCRC := flags&SMFlagCRC64 != 0
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azdatalake/internal/shared/structured_message.go:209]: Same reasoning as the azfile copy above.

**Review by souravgupta-msft — APPROVED** [review]

**Review by souravgupta-msft — APPROVED** [review]
